### PR TITLE
query vc realtime data

### PR DESCRIPTION
**Pull request by VMware-HaoyuLi**
Signed-off-by: haoyul <haoyul@vmware.com>

### CODE REVIEW

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 1]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+      MetricId[] metricIds = metricIdList.toArray(new MetricId[metricIdList.size()]);
+
+      ProviderSummary summary = performanceManager.queryProviderSummary(host._getRef());
+      int perfInterval = summary.getRefreshRate();
+      QuerySpec[] specs = new QuerySpecImpl[1];
+      specs[0] = new QuerySpecImpl();
+      specs[0].setEntity(host._getRef());
+      specs[0].setMetricId(metricIds);
+      specs[0].setIntervalId(perfInterval);
+      specs[0].setMaxSample(1);
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:352]: double check if we can query the last 5minutes.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 2]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+               && unitInfoSummary.equalsIgnoreCase(counterInfo.getUnitInfo().getSummary())
+               && nameInfoSummary.equalsIgnoreCase(counterInfo.getNameInfo().getSummary())) {
+            counters.put(new Integer(counterInfo.getKey()), counterInfo);
+         }
+      }
+      MetricId[] queryAvailableMetric =
+            performanceManager.queryAvailableMetric(host._getRef(), null, null, new Integer(20));
+      ArrayList<MetricId> metricIdList = new ArrayList<MetricId>();
+      if (queryAvailableMetric.length > 0) {
+         for (int i = 0; i < queryAvailableMetric.length; i++) {
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:335]: check the metric id format.

**Review by yixingjia — CHANGES_REQUESTED** [review]

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java (hunk 2 above)
Comment (VMware-HaoyuLi) [reply vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:335]: (vim.MetricId) {
   dynamicType = null,
   dynamicProperty = null,
   counterId = 182,
   instance = d7c13f5a-90cbd366
}

**Comment by yixingjia** [comment]
@YangJiao0817  Please help review this PR.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/client/VsphereClient.java [hunk 3]
```diff
@@ -20,8 +20,10 @@
 
 import com.vmware.flowgate.vcworker.model.VCConstants;
 import com.vmware.vim.binding.impl.vmodl.TypeNameImpl;
+import com.vmware.vim.binding.vim.ClusterComputeResource;
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/client/VsphereClient.java:23]: Useless import.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 4]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+      return VsphereClient.connect(String.format(VCConstants.SDKURL, vc.getServerURL()),
+            vc.getUserName(), vc.getPassword(), !vc.isVerifyCert());
+   }
+
+   public HashMap<String, ServerMapping> getVaildServerMapping(SDDCSoftwareConfig vc) {
+
+      HashMap<String, ServerMapping> mobIdDictionary = new HashMap<String, ServerMapping>();
+      ServerMapping[] mappings = null;
+      try {
+         restClient.setServiceKey(serviceKeyConfig.getServiceKey());
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:220]: restClient.setServiceKey repeat

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 5]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+         mappings = restClient.getServerMappingsByVC(vc.getId()).getBody();
+      } catch (HttpClientErrorException clientError) {
+         if (clientError.getRawStatusCode() != HttpStatus.NOT_FOUND.value()) {
+            return null;
+         }
+      }
+
+      for (ServerMapping mapping : mappings) {
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:228]: Check mappings is null

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 6]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+            mobIdDictionary.put(mapping.getVcMobID(), mapping);
+         }
+      }
+      return mobIdDictionary;
+
+   }
+   public void queryHostMetrics(SDDCSoftwareConfig vc) {
+
+      HashMap<String, ServerMapping> serverMappingMap = getVaildServerMapping(vc);
+      if (serverMappingMap == null) {
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:239]: serverMappingMap cannot be null

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 7]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+         logger.info("serverMapping is invaild");
+         return;
+      }
+      try (VsphereClient vsphereClient = connectVsphere(vc);) {
+
+         Collection<HostSystem> hosts = vsphereClient.getAllHost();
+         if (hosts == null || hosts.isEmpty()) {
+            logger.error("vsphere: " + vsphereClient.getVCUUID() + " get hosts is null");
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:247]: logger.error("vsphere: {} get hosts is null", vsphereClient.getVCUUID());

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 8]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+         for (HostSystem host : hosts) {
+
+            String mobId = host._getRef().getValue();
+            if (serverMappingMap.containsKey(mobId)) {
+               ServerMapping serverMapping = serverMappingMap.get(mobId);
+               String assetId = serverMapping.getAsset();
+
+               Asset hostMappingAsset = restClient.getAssetByID(assetId).getBody();
+               if (hostMappingAsset == null) {
+                  logger.error("serverMapping: " + serverMapping.getId() + " get asset: " + assetId + " is null");
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:259]: logger.error("serverMapping:{} get asset: {} is null", serverMapping.getId(), assetId);

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 9]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
    }
 
+   public VsphereClient connectVsphere(SDDCSoftwareConfig vc) throws Exception {
+      return VsphereClient.connect(String.format(VCConstants.SDKURL, vc.getServerURL()),
+            vc.getUserName(), vc.getPassword(), !vc.isVerifyCert());
+   }
+
+   public HashMap<String, ServerMapping> getVaildServerMapping(SDDCSoftwareConfig vc) {
+
+      HashMap<String, ServerMapping> mobIdDictionary = new HashMap<String, ServerMapping>();
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:217]: Map<String, ServerMapping> mobIdDictionary = new HashMap<String, ServerMapping>();

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 10]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+
+         if (groupInfoKey.equalsIgnoreCase(counterInfo.getGroupInfo().getKey())
+               && unitInfoSummary.equalsIgnoreCase(counterInfo.getUnitInfo().getSummary())
+               && nameInfoSummary.equalsIgnoreCase(counterInfo.getNameInfo().getSummary())) {
+            counters.put(new Integer(counterInfo.getKey()), counterInfo);
+         }
+      }
+      MetricId[] queryAvailableMetric =
+            performanceManager.queryAvailableMetric(host._getRef(), null, null, new Integer(20));
+      ArrayList<MetricId> metricIdList = new ArrayList<MetricId>();
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:333]: List<MetricId> metricIdList = new ArrayList<MetricId>();

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/client/VsphereClient.java [hunk 11]
```diff
@@ -122,6 +125,9 @@ public void onSuccess(X509Certificate[] chain, String thumbprint, Result verifyR
       ServiceInstance si = client.createStub(ServiceInstance.class, svcRef);
 
       sic = si.retrieveContent();
+      
+      performanceManager = client.createStub(PerformanceManager.class, sic.getPerfManager());
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/client/VsphereClient.java:129]: Lazy loading

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 12]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+
+      List<RealTimeData> realTimeDatas = new ArrayList<RealTimeData>();
+      RealTimeData realTimeData = new RealTimeData();
+      realTimeData.setAssetID(hostMappingAsset.getId());
+      List<ValueUnit> values = new ArrayList<ValueUnit>();
+
+      PerformanceManager performanceManager = vsphereClient.getPerformanceManager();
+      CounterInfo[] counterInfos = performanceManager.getPerfCounter();
+
+      Map<String, Long> cpuUsage = getRealTimeDataByType(performanceManager, counterInfos, host,
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:397]: Parameter extraction cycle call

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 13]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+      storageIORateUsageValue.setKey(MetricName.VC_HOST_STORAGEIORATEUSAGE);
+      storageIORateUsageValue.setTime(storageIORateUsage.get(VCConstants.TIMESTAMP));
+      storageIORateUsageValue.setValueNum(storageIORateUsage.get(VCConstants.VALUE));
+      storageIORateUsageValue.setUnit(RealtimeDataUnit.Kbps.toString());
+      values.add(storageIORateUsageValue);
+
+      long storageCapacity = 0;
+      long storageUsed = 0;
+      DatastoreInfo[] datastores = host.queryConnectionInfo().getDatastore();
+      for (DatastoreInfo datastore : datastores) {
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:492]: check null

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 14]
```diff
@@ -168,6 +207,315 @@ public void checkAndUpdateIntegrationStatus(SDDCSoftwareConfig vc, String messag
...
+      storageUsedValue.setKey(MetricName.VC_HOST_STORAGEUSED);
+      storageUsedValue.setTime(cpuUsage.get(VCConstants.TIMESTAMP));
+      storageUsedValue.setValueNum(storageCapacity);
+      storageUsedValue.setUnit(RealtimeDataUnit.Mb.toString());
+      values.add(storageUsedValue);
+
+      ValueUnit storageUsageValue = new ValueUnit();
+      storageUsageValue.setKey(MetricName.VC_HOST_STORAGEUSAGE);
+      storageUsageValue.setTime(cpuUsage.get(VCConstants.TIMESTAMP));
+      storageUsageValue.setValueNum((double) (storageUsed * 100) / storageCapacity);
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:508]: If it is accurate transportation, use BigDecimal

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java (hunk 4 above)
Comment (VMware-HaoyuLi) [reply vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:220]: It is necessary.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/client/VsphereClient.java [hunk 15]
```diff
@@ -206,6 +211,10 @@ public ServiceInstanceContent getServiceInstanceContent() {
       return sic;
    }
 
+   public PerformanceManager getPerformanceManager() {
+      return client.createStub(PerformanceManager.class, sic.getPerfManager());
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/client/VsphereClient.java:215]: Need refactor, only create one stub is enough.

**Review by yixingjia — CHANGES_REQUESTED** [review]

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 16]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
       }
    }
 
-   public boolean feedClusterMetaData(HashMap<String, ClusterComputeResource> clusterMap,
+   public void queryHostMetrics(SDDCSoftwareConfig vc) {
+
+      Map<String, ServerMapping> serverMappingMap = getVaildServerMapping(vc);
+      if (serverMappingMap == null || serverMappingMap.isEmpty()) {
+         logger.info("serverMapping is invaild");
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:341]: No serverMapping found for vc xxx

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 17]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+         return;
+      }
+      try (VsphereClient vsphereClient = connectVsphere(vc);) {
+
+         Collection<HostSystem> hosts = vsphereClient.getAllHost();
+         if (hosts == null || hosts.isEmpty()) {
+            logger.error("vsphere: {} get hosts is null", vsphereClient.getVCUUID());
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:348]: vc Name or host ip address.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 18]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+         for (HostSystem host : hosts) {
+
+            String mobId = host._getRef().getValue();
+            if (serverMappingMap.containsKey(mobId)) {
+               ServerMapping serverMapping = serverMappingMap.get(mobId);
+               String assetId = serverMapping.getAsset();
+
+               Asset hostMappingAsset = restClient.getAssetByID(assetId).getBody();
+               if (hostMappingAsset == null) {
+                  logger.error("serverMapping: {} get asset: {} is null", serverMapping.getId(), assetId);
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:360]: The asset {} doesn't exist in serverMapping {}.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 19]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+               feedHostUsageData(vsphereClient, assetId, host._getRef());
+            }
+         }
+
+      } catch (ConnectionException e1) {
+         checkAndUpdateIntegrationStatus(vc, e1.getMessage());
+         return;
+      } catch (ExecutionException e2) {
+         if (e2.getCause() instanceof InvalidLogin) {
+            logger.error("Failed to push data to {} {}" ,vc.getServerURL(), e2);
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:372]: Correct the logging message.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 20]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+            checkAndUpdateIntegrationStatus(vc, "Invalid username or password.");
+            return;
+         }
+      } catch (Exception e) {
+         logger.error("Failed to sync the host metadata to VC ", e);
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:377]: logging message.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 21]
```diff
@@ -812,7 +1082,20 @@ private void syncCustomerAttrsData(SDDCSoftwareConfig vcInfo) {
...
+                        if (null != metricsFormulars) {
+                           Map<String, Map<String, String>> metrics =
+                                 metricsFormulars.get(FlowgateConstant.HOST_METRICS);
+                           if (metrics == null || metrics.isEmpty()) {
+                              metrics = new HashMap<>();
+                              feedAssetMetricsFormulars(metrics, assetId);
+                              metricsFormulars.put(FlowgateConstant.HOST_METRICS, metrics);
+                              restClient.saveAssets(asset);
+                           }
+                        }
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:1098]: refactor into a method . Please also check when the mapping already exist situation.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 22]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+         ManagedObjectReference hostRef) {
+
+      List<RealTimeData> realTimeDatas = new ArrayList<RealTimeData>();
+      RealTimeData realTimeData = new RealTimeData();
+      realTimeData.setAssetID(assetId);
+      List<ValueUnit> valueUnits = new ArrayList<ValueUnit>();
+ 
+      PerformanceManager performanceManager = vsphereClient.getPerformanceManager();
+
+      Map<Integer, String> counters = filterVaildCounters(performanceManager);
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:489]: getMetricsCounters()

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 23]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+      if(counters == null || counters.isEmpty()) {
+         logger.error("Asset: {} failed to get performance counters", assetId);
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:491]: metrics

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 24]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+         return;
+      }
+      List<MetricId> metricIdList = filterVaildMetricId(performanceManager, hostRef, counters);
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:494]: getPerformenceMetricsIds

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 25]
```diff
@@ -325,7 +334,268 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+      specs[0].setMaxSample(15);
+
+      //get metrics value
+      EntityMetricBase[] metricBase = performanceManager.queryStats(specs);
+      
+      for (int i = 0; metricBase != null && i < metricBase.length; i++) {
+         EntityMetricBase entityMetricBase = metricBase[i];
+         EntityMetric entityMetric = (EntityMetric) entityMetricBase;
+         MetricSeries[] metricSeries = entityMetric.getValue();
+         SampleInfo[] sampleInfos = entityMetric.getSampleInfo();
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:516]: double check.

**Review by yixingjia — CHANGES_REQUESTED** [review]

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 26]
```diff
@@ -250,7 +259,7 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
             return;
          }
          Collection<ClusterComputeResource> clusters = vsphereClient.getAllClusterComputeResource();
-         HashMap<String, ClusterComputeResource> clusterMap =
+         Map<String, ClusterComputeResource> clusterMap =
                new HashMap<String, ClusterComputeResource>();
          for (ClusterComputeResource cluster : clusters) {
```
Comment (YangJiao0817) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:264]: Check if clusters is null.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 27]
```diff
@@ -325,7 +338,280 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+         return;
+      }
+      for(EntityMetricBase entityMetricBase : metricBase) {
+         EntityMetric entityMetric = (EntityMetric) entityMetricBase;
+         MetricSeries[] metricSeries = entityMetric.getValue();
+         SampleInfo[] sampleInfos = entityMetric.getSampleInfo();
+         for (MetricSeries metricSerie : metricSeries) {
+            if (metricSerie instanceof IntSeries) {
+               IntSeries intSeries = (IntSeries) metricSerie;
+               long[] values = intSeries.getValue();
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:538]: please check what will happen if it don't have 15 data point.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 28]
```diff
@@ -325,7 +338,280 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+               int counterId = metricSerie.getId().getCounterId();
+               for(int index = 0; index < values.length; ++index) {
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:540]: index++

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 29]
```diff
@@ -325,7 +338,280 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+                  long timeStamp = sampleInfos[index].getTimestamp().getTimeInMillis();
+                  
+                  ValueUnit valueUnit = new ValueUnit();
+                  long v = values[index];
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:544]: long name

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java (hunk 27 above)
Comment (VMware-HaoyuLi) [reply vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:538]: The size of the array must match the size of sampleInfo property in the PerfEntityMetric that contains this series.

File: vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java [hunk 30]
```diff
@@ -310,22 +323,358 @@ public void queryHostMetaData(SDDCSoftwareConfig vc) {
...
+                  valueUnit.setTime(timeStamp);
+                  valueUnits.add(valueUnit);
+               }
+            }
+         }
+         
+      }
+
+      realTimeData.setValues(valueUnits);
+      realTimeData.setTime(valueUnits.get(0).getTime());
```
Comment (yixingjia) [inline vc-worker/src/main/java/com/vmware/flowgate/vcworker/scheduler/job/VCDataService.java:670]: what if the valueUnits is empty.

**Review by yixingjia — APPROVED** [review]

**Pull request merged**